src/modules/mavlink/mavlink_sign_control.cpp

```diff
@@ -0,0 +1,214 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mavlink_sign_control.cpp
+ * Mavlink messages signing control helpers implementation.
+ *
+ * @author Yulian oifa <yulian.oifa@mobius-software.com>
+ */
+
+#include "mavlink_sign_control.h"
+#include <sys/stat.h>
+
+static mavlink_signing_streams_t global_mavlink_signing_streams = {};
+
+static const uint32_t unsigned_messages[] = {
+	MAVLINK_MSG_ID_RADIO_STATUS,
+	MAVLINK_MSG_ID_ADSB_VEHICLE,
+	MAVLINK_MSG_ID_COLLISION
+};
+
+MavlinkSignControl::MavlinkSignControl()
+{
+}
+
+MavlinkSignControl::~MavlinkSignControl()
+{
+}
+
+void MavlinkSignControl::start(int _instance_id, mavlink_status_t *_mavlink_status,
+			       mavlink_accept_unsigned_t accept_unsigned_callback)
+{
+	_mavlink_signing.link_id = _instance_id;
+	_mavlink_signing.flags = MAVLINK_SIGNING_FLAG_SIGN_OUTGOING;
+	_mavlink_signing.accept_unsigned_callback = accept_unsigned_callback;
+	is_signing_initialized = false;
+
+	int mkdir_ret = mkdir(MAVLINK_FOLDER_PATH, S_IRWXU);
+
+	if (mkdir_ret != 0 && errno != EEXIST) {
+		PX4_ERR("failed creating module storage dir: %s (%i)", MAVLINK_FOLDER_PATH, errno);
+
+	} else {
+		int _fd = ::open(MAVLINK_SECRET_FILE, O_CREAT | O_RDONLY, PX4_O_MODE_600);
+
+		if (_fd == -1) {
+			if (errno != ENOENT) {
+				PX4_ERR("failed creating mavlink secret key file: %s (%i)", MAVLINK_SECRET_FILE, errno);
+			}
+
+		} else {
+			//if we dont have enough bytes we simply ignore it , because it may be not set yet
+			ssize_t bytes_read = ::read(_fd, _mavlink_signing.secret_key, 32);
+
+			if (bytes_read == 32) {
+				bytes_read = ::read(_fd, &_mavlink_signing.timestamp, 8);
+
+				if (bytes_read == 8) {
+					if (_mavlink_signing.timestamp != 0 || !is_array_all_zeros(_mavlink_signing.secret_key, 32)) {
+						is_signing_initialized = true;
+					}
+				}
+			}
+
+			close(_fd);
+		}
+	}
+
+	//lets reset it to nulls if it was not read properly
+	if (!is_signing_initialized) {
+		for (size_t i = 0; i < 32; ++i) {
+			_mavlink_signing.secret_key[i] = 0;
+		}
+
+		_mavlink_signing.timestamp = 0;
+	}
+
+	// copy pointer of the signing to status struct
+	_mavlink_status -> signing = &_mavlink_signing;
+	_mavlink_status -> signing_streams = &global_mavlink_signing_streams;
+}
+
+bool MavlinkSignControl::check_for_signing(const mavlink_message_t *msg)
+{
+	if (msg->msgid == MAVLINK_MSG_ID_SETUP_SIGNING) {
```

Comment:
nit pick, but we should prefer the return early pattern. If you choose to do this, apply the pattern on all of the new code.